### DISCUSSION TITLE
How to use RouterContext ?

### DISCUSSION THREAD

**Discussion by carragom** [oakserver/oak#547] [Q&A]
Hello,
Consider the following code
```typescript
// test.ts
import { Application, Router, RouterContext } from "https://deno.land/x/oak@v11.0.0/mod.ts";

const router = new Router();
router.get('/test/:myparam', testHandler);

const app = new Application();
app.use(router.allowedMethods());
app.use(router.routes());
await app.listen({port: 3000});

function testHandler(ctx: RouterContext) {
	console.debug(ctx.params);
	ctx.response.body = `Path parameter was '${ctx.params.myparam}'`;
}
```
When run with `deno run --allow-net test.ts` and accessed with `curl localhost:3000/test/asd` will respond with `Path parameter was 'asd'` as expected.

But when you type check that with `deno run --allow-net --check test.ts` or simply open it with vsode and the deno extension, it throws the following: `error: TS2707 [ERROR]: Generic type 'RouterContext<R, P, S>' requires between 1 and 3 type arguments`. 

What would be the right type definition for `ctx` in the `testHandler` function to make Typescript happy?.
Thanks.

**Comment by lukepuplett** [comment]
I'm just learning too. Is it just `Context`?

**Comment by CwhyNK** [comment]
For anyone still finding an answer: `RouterContext` is a generic interface so you will need to supply at least one type parameter when declaring it; like this `RouterContext<"/:myparam">`

**Reply by martin-braun** [reply]
@kitsonk Deep-searching the web helped me to find this. Oak definitely needs some proper docs, the JSR docs don't do justice.

**Comment by carragom** [comment]
The requirement I had when I tried to use RouterContext was getting access to route parameters. Since I found a different way to get access to route parameters and the RouterContext API is quite uncomfortable to be honest I opted for not using RouterContext at all. This is what I ended up with.

```typescript
import { Context, helpers } from '@oak/oak';

function testHandler(ctx: Context) {
        const params = helpers.getQuery(ctx, { mergeParams: true });
	console.debug(params);
	ctx.response.body = `Path parameter was '${params.myparam}'`;
}
```

See the [docs](https://jsr.io/@oak/oak/doc/helpers/~/getQuery) for more
